src/scalib_ext/scalib/src/snr.rs

```diff
@@ -93,193 +245,305 @@ impl SNR {
+///  all values in y must be < sum.shape()[1]
+unsafe fn inner_snr_update(
+    // len: n
+    trace_chunk: &[[i16; 8]],
+    // (np, n)
+    y: ArrayView2<u16>,
+    // (np, nc)
+    mut sum: ArrayViewMut2<[i32; 8]>,
+    sum_square: &mut [i64; 8],
+) {
+    assert_eq!(trace_chunk.len(), y.shape()[1]);
+    assert_eq!(sum.shape()[0], y.shape()[0]);
+    for trace in trace_chunk {
+        for (sum_square, trace) in sum_square.iter_mut().zip(trace.iter()) {
+            let trace = *trace as i64;
+            // overflow handled with error elsewhere
+            *sum_square = sum_square.wrapping_add(trace * trace);
+        }
+    }
+    izip!(y.outer_iter(), sum.outer_iter_mut()).for_each(|(y, sum)| {
+        let sum = sum.into_slice().unwrap();
+        izip!(y.to_slice().unwrap(), trace_chunk).for_each(|(y, trace_chunk)| {
+            // sum.get_unchecked_mut is safe due to assumption,
+            let sum = sum.get_unchecked_mut(*y as usize);
+            for j in 0..8 {
+                // overflow handled with error elsewhere
+                sum[j] = sum[j].wrapping_add(trace_chunk[j] as i32);
+            }
+        })
+    });
 }
 
-#[inline(always)]
-fn inner_loop_update(
-    sum: ArrayViewMut1<i64>,
-    sum_square: ArrayViewMut1<i64>,
-    x: ArrayView2<i16>,
-    y: ArrayView1<u16>,
-    i: u16,
-) {
-    let sum = sum.into_slice().unwrap();
-    let sum_square = sum_square.into_slice().unwrap();
-    izip!(x.outer_iter(), y.iter()).for_each(|(x, v)| {
-        if i == *v {
-            let x = x.to_slice().unwrap();
-            izip!(sum.iter_mut(), sum_square.iter_mut(), x.iter()).for_each(
-                |(sum, sum_square, &x)| {
-                    let x = x as i64;
-                    *sum += x;
-                    *sum_square += x * x;
-                },
-            );
+#[inline(never)]
+fn transpose_traces(
+    // shape: (4, n)
+    mut traces_tr: ArrayViewMut2<[i16; 8]>,
+    // shape: (n, ns) with ns <= 32
+    trace_chunk: ArrayView2<i16>,
+) -> i16 {
+    assert_eq!(traces_tr.shape()[1], trace_chunk.shape()[0]);
+    assert_eq!(traces_tr.shape()[0], 4);
+    assert!(trace_chunk.shape()[1] <= 32);
+    let mut max_width: i16 = 0;
+    if trace_chunk.shape()[1] == 32 {
+        let mut max_width_vec = [0i16; 8];
+        izip!(
+            traces_tr.axis_iter_mut(Axis(1)),
+            trace_chunk.axis_iter(Axis(0))
+        )
+        .for_each(|(mut traces_tr, trace_chunk)| {
+            izip!(
+                traces_tr.iter_mut(),
+                trace_chunk.axis_chunks_iter(Axis(0), 8)
+            )
+            .for_each(|(traces_tr, trace_chunk)| {
+                let trace_chunk: &[i16; 8] = trace_chunk.to_slice().unwrap().try_into().unwrap();
+                //traces_tr.clone_from_slice(trace_chunk);
+                *traces_tr = *trace_chunk;
+                for (max_width, trace_chunk) in max_width_vec.iter_mut().zip(trace_chunk.iter()) {
+                    *max_width |= trace_chunk.abs();
```

Comment:
We cannot abs on i16::MIN https://doc.rust-lang.org/std/primitive.i16.html#method.abs
This panics! if full 16-bit precision is used.

Reply:
yep, fixed